server/src/test-fast/java/com/thoughtworks/go/server/service/GoDashboardServiceTest.java

```diff
@@ -178,49 +177,127 @@ public void allPipelineGroupsForDashboard_shouldRetrievePipelineGroupsBasedOnUse
 
-        when(filter.isPipelineVisible(any(CaseInsensitiveString.class))).thenReturn(true);
+        assertThat(envs.get(1).pipelines(), contains("pipeline3"));
+    }
+
+    @Test
+    public void allPipelineGroupsForDashboard_shouldNotListEmptyPipelineGroup() {
+        DashboardFilter filter = new WhitelistFilter("nothing", Collections.emptyList());
+        configMother.addPipelineWithGroup(config, "group1", "pipeline1", "stage1A", "job1A1");
+        addPipelinesToCache(pipeline("pipeline1", "group1"));
 
         List<GoDashboardPipelineGroup> pipelineGroups = allPipelineGroupsForDashboard(filter, new Username("user1"));
 
         assertThat(pipelineGroups.size(), is(0));
     }
 
     @Test
-    public void allPipelineGroupsForDashboard_shouldNotListPipelinesExistingInConfigButNotInCache() throws Exception {
-        DashboardFilter filter = mock(DashboardFilter.class);
+    public void allEnvironmentsForDashboard_shouldNotListEmptyPipelineGroup() {
+        DashboardFilter filter = new WhitelistFilter("nothing", Collections.emptyList());
+        configMother.addPipelineWithGroup(config, "group1", "pipeline1", "stage1A", "job1A1");
+        addPipelinesToCache(pipeline("pipeline1", "group1"));
 
+        configMother.addEnvironmentConfig(config, "env1", "pipeline1");
+        List<GoDashboardEnvironment> pipelineGroups = allEnvironmentsForDashboard(filter, new Username("user1"));
+
+        assertThat(pipelineGroups.size(), is(0));
+    }
+
+    @Test
+    public void allPipelineGroupsForDashboard_shouldNotListPipelinesExistingInConfigButNotInCache() {
         configMother.addPipelineWithGroup(config, "group1", "pipeline2", "stage1A", "job1A1");
 
         configMother.addPipelineWithGroup(config, "group1", "pipeline1", "stage1A", "job1A1");
         GoDashboardPipeline pipeline1 = pipeline("pipeline1", "group1", new Permissions(Everyone.INSTANCE,
                 Everyone.INSTANCE, Everyone.INSTANCE, Everyone.INSTANCE));
 
         addPipelinesToCache(pipeline1);
-        when(filter.isPipelineVisible(any(CaseInsensitiveString.class))).thenReturn(true);
 
-        List<GoDashboardPipelineGroup> pipelineGroups = allPipelineGroupsForDashboard(filter, new Username("user1"));
+        List<GoDashboardPipelineGroup> pipelineGroups = allPipelineGroupsForDashboard(Filters.WILDCARD_FILTER, new Username("user1"));
 
         assertThat(pipelineGroups.size(), is(1));
-        assertThat(pipelineGroups.get(0).allPipelineNames(), contains("pipeline1"));
-        assertThat(pipelineGroups.get(0).allPipelines(), contains(pipeline1));
+        assertThat(pipelineGroups.get(0).pipelines(), contains("pipeline1"));
+        assertThat(pipelineGroups.get(0).pipelines(), not(contains("pipeline2")));
     }
 
     @Test
-    public void shouldRemoveExistingPipelineEntryInCacheWhenPipelineConfigIsRemoved() throws Exception {
+    public void allEnvironmentsForDashboard_shouldNotListPipelinesExistingInConfigButNotInCache() {
```

Comment:
@akshaydewan yes, ideally; however I hadn't thought enough about it to know whether or not it's every possible for it to be out of sync. If it has, nobody has noticed, certainly. I would not be against removing it.

Reply:
Let's remove it then.

Reply:
@akshaydewan I dug into the history, and it looks like this test was added as part of 0886bae1c2f59bcc3e88f661f6cc37d1789f7faf

The commit indicates @maheshp and @jyotisingh paired on this. Perhaps ask Mahesh for more context? I'm sure they wouldn't have added this without good reason, but it may be too subtle for me.